contracts/staking/GoodFundManager.sol

```diff
@@ -290,66 +291,72 @@ contract GoodFundManager is DAOUpgradeableContract, DSMath {
 		uint256 tempInterest;
 		int256 i;
-		for (i = 0; i < int256(activeContractsLength); i++) {
+		for (i = 0; i < int256(activeContracts.length); i++) {
 			(, , , , tempInterest) = IGoodStaking(activeContracts[uint256(i)])
 				.currentGains(false, true);
 			if (tempInterest != 0) {
 				addresses[uint256(i)] = activeContracts[uint256(i)];
 				balances[uint256(i)] = tempInterest;
 			}
 		}
-		uint256 gasCostInDAI = getGasPriceIncDAIorDAI(_maxGasAmount, true); // Get gas price in DAI so can compare with possible interest amount to get
-		address[] memory emptyArray = new address[](0);
-
+		uint256 usedGasAmount = gasCostExceptInterestCollect;
 		quick(balances, addresses); // sort the values according to interest balance
 		uint256 gasCost;
 		uint256 possibleCollected;
-		for (i = int256(activeContractsLength) - 1; i >= 0; i--) {
+		for (i = int256(activeContracts.length) - 1; i >= 0; i--) {
 			// elements are sorted by balances from lowest to highest
 
 			if (addresses[uint256(i)] != address(0x0)) {
 				gasCost = IGoodStaking(addresses[uint256(i)])
 					.getGasCostForInterestTransfer();
-				if (_maxGasAmount - gasCost >= gasCostExceptInterestCollect) {
-					// collects the interest from the staking contract and transfer it directly to the reserve contract
-					//`collectUBIInterest` returns (iTokengains, tokengains, precission loss, donation ratio)
-					possibleCollected += balances[uint256(i)];
-					_maxGasAmount = _maxGasAmount - gasCost;
-				} else {
-					break;
-				}
+
+				// collects the interest from the staking contract and transfer it directly to the reserve contract
+				//`collectUBIInterest` returns (iTokengains, tokengains, precission loss, donation ratio)
+				possibleCollected += balances[uint256(i)];
+				usedGasAmount += gasCost;
+				gasCostsSoFar[uint256(i)] = usedGasAmount;
+				collectedInterestSoFar[uint256(i)] = possibleCollected;
 			} else {
 				break; // if addresses are null after this element then break because we initialize array in size activecontracts but if their interest balance is zero then we dont put it in this array
 			}
 		}
-		while (i > -1) {
-			addresses[uint256(i)] = address(0x0);
-			i -= 1;
-		}
-		if (
-			block.timestamp >=
+		uint256 maxGasAmount = block.timestamp >=
 			lastCollectedInterest + collectInterestTimeThreshold
-		) {
-			if (possibleCollected * 1e10 < gasCostInDAI) return emptyArray; // multiply possiblecollected by 1e10 so it will be on the 18decimals
-		} else {
-			// multiply possiblecollected by 1e10 so it will be on the 18decimals
-			if (possibleCollected * 1e10 < interestMultiplier * gasCostInDAI)
-				return emptyArray;
-		}
-		return addresses;
+			? (possibleCollected * 1e10) / getGasPriceIncDAIorDAI(1, true)
+			: (possibleCollected * 1e10) /
+				(interestMultiplier * getGasPriceIncDAIorDAI(1, true));
+		return (
+			addresses,
+			balances,
+			collectedInterestSoFar,
+			gasCostsSoFar,
+			maxGasAmount,
+			maxGasAmount >= usedGasAmount
```

Comment:
everything here should be a PER ROW data. this is a single boolean

Reply:
Ok I'll move it to struct